code/modules/cargo/expressconsole.dm

```diff
@@ -68,29 +90,57 @@
 		ui = new(user, src, ui_key, "cargo_express", name, 1000, 800, master_ui, state)
 		ui.open()
 
-
 /obj/machinery/computer/cargo/express/ui_data(mob/user)
+	var/canBeacon = beacon && (isturf(beacon.loc) || ismob(beacon.loc))
 	var/list/data = list()
 	data["locked"] = locked
 	data["siliconUser"] = user.has_unlimited_silicon_privilege
+	data["beaconzone"] = beacon ? get_area(beacon) : ""
+	data["usingBeacon"] = usingBeacon
+	data["canBeacon"] = !usingBeacon || canBeacon
+	data["canBuyBeacon"] = cooldown <= 0 && SSshuttle.points >= BEACON_COST
+	data["beaconError"] = usingBeacon && !canBeacon ? "(BEACON ERROR)" : ""
+	data["hasBeacon"] = beacon != null
+	data["beaconName"] = beacon ? beacon.name : "No Beacon Found"
+	data["printMsg"] = cooldown > 0 ? "Print Beacon for [BEACON_COST] credits ([cooldown])" : "Print Beacon for [BEACON_COST] credits"
 	data["points"] = SSshuttle.points
 	data["supplies"] = list()
 	message = "Sales are near-instantaneous - please choose carefully."
 	if(SSshuttle.supplyBlocked)
 		message = blockade_warning
+	if(usingBeacon && !beacon)
+		message = "BEACON ERROR: BEACON MISSING"//beacon was destroyed
+	else if (usingBeacon && !canBeacon)
+		message = "BEACON ERROR: MUST BE EXPOSED"//beacon's loc/user's loc must be a turf
 	if(obj_flags & EMAGGED)
 		message = "(&!#@ERROR: ROUTING_#PROTOCOL MALF(*CT#ON. $UG%ESTE@ ACT#0N: !^/PULS3-%E)ET CIR*)ITB%ARD."
-
 	data["message"] = message
 	if(!meme_pack_data)
 		packin_up()
 		stack_trace("You didn't give the cargo tech good advice, and he ripped the manifest. As a result, there was no pack data for [src]")
 	data["supplies"] = meme_pack_data
-
+	if (cooldown > 0)
+		cooldown--
 	return data
 
 /obj/machinery/computer/cargo/express/ui_act(action, params, datum/tgui/ui)
 	switch(action)
+		if("LZCargo")
+			usingBeacon = FALSE
+			if (beacon)
+				beacon.update_status(SP_UNREADY) //ready light on beacon will turn off
+		if("LZBeacon")
+			usingBeacon = TRUE
+			if (beacon)
+				beacon.update_status(SP_READY) //turns on the beacon's ready light
+		if("printBeacon")
+			cooldown = 10//a ~ten second cooldown for printing beacons to prevent spam
+			var/obj/item/supplypod_beacon/C = new /obj/item/supplypod_beacon(drop_location())
+			C.link_console(src, usr)//rather than in beacon's Initialize(), we can assign the computer to the beacon by reusing this proc)
+			printed_beacons++//printed_beacons starts at 0, so the first one out will be called beacon # 1
+			beacon.name = "Supply Pod Beacon #[printed_beacons]"
+			SSshuttle.points -= BEACON_COST//check for this is under ui_data

```

Comment:
UI checks can't be fully trusted, someone might spamclick this button multiple times before the UI updates. You still need a conditional here.